specs/rollup-node-p2p.md

```diff
@@ -0,0 +1,258 @@
+- `D_low` (topic stable mesh low watermark): 6
+- `D_high` (topic stable mesh high watermark): 12
+- `D_lazy` (gossip target): 6
+- `heartbeat_interval` (frequency of heartbeat, seconds): 0.5
+- `fanout_ttl` (ttl for fanout maps for topics we are not subscribed to but have published to, seconds): 24
+- `mcache_len` (number of windows to retain full messages in cache for `IWANT` responses): 12
+- `mcache_gossip` (number of windows to gossip about): 3
+- `seen_ttl` (number of heartbeat intervals to retain message IDs): 40
+
+Notable differences from L1 consensus (Eth2):
+
+- `seen_ttl` does not need to cover a full L1 epoch (6.4 minutes), but rather just a small window covering latest blocks
+- `fanout_ttl`: adjusted to lower than `seen_ttl`
+- `mcache_len`: a larger number of heartbeats can be retained since the gossip is much less noisy.
+- `heartbeat_interval`: faster interval to reduce latency, bandwidth should still be reasonable since
+  there are far fewer messages to gossip about each interval than on L1 which uses an interval of 0.7 seconds.
+
+#### Topic configuration
+
+Topics have string identifiers and are communicated with messages and subscriptions.
+`/optimism/chain_id/hardfork_version/Name`
+
+- `chain_id`: replace with decimal representation of chain ID
+- `hardfork_version`: replace with decimal representation of hardfork
+- `Name`: topic application-name
+
+Note that the topic encoding depends on the topic, unlike L1,
+since there are less topics, and all are snappy-compressed.
+
+## Gossip Topics
+
+### `blocks`
+
+The primary topic of the L2, to distribute blocks to other nodes faster than proxying through L1 would.
+
+#### Block encoding
+
+TODO: encode execution payload (SSZ or RLP), with sequencer identifier and signature.
+
+TODO: signature type and verification (`secp256k1` like transactions, with some different signature domain)
+
+#### Block validation
+
+To ensure malicious peers get scored based on application behavior the validation signals
+`ACCEPT` (valid), `IGNORE` (like inactivity) or `REJECT` (score penalties).
+
+In order of operation:
+
+- `[REJECT]` if the encoding or compression is not valid
+- `[REJECT]` if the block timestamp is older than 20 seconds in the past
+  (graceful boundary for worst-case propagation and clock skew)
+- `[REJECT]` if the block timestamp is more than 5 seconds into the future (graceful boundary for clock skew)
+- `[REJECT]` if the signature by the sequencer is not valid
+- `[REJECT]` if more than 5 blocks have been seen with the same block height
+- `[IGNORE]` if a conflicting block was already seen on L1.
+  It may not be malicious due to racing between L1 confirmation and L2 propagation, but should be filtered out.
+
+The block is signed by the corresponding sequencer, to filter malicious messages.
+The sequencer model is singular but may change to multiple sequencers in the future.
+A default sequencer pubkey is distributed with rollup nodes and should be configurable.
```

Comment:
If we decide to run clique consensus, we likely want multiple nodes using different keys for redundancy. If the key is hardcoded, then we will have a really bad time if the key is leaked or lost

Reply:
Yep, need some key-replacement mechanism. We can discuss what's the best approach later, for now I'm going to keep it simple.